torch/csrc/utils/tensor_numpy.cpp

```diff
@@ -198,6 +202,102 @@ bool is_numpy_scalar(PyObject* obj) {
           PyArray_IsScalar(obj, Floating));
 }
 
+at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+  PyObject *cuda_dict = PyObject_GetAttrString(obj, "__cuda_array_interface__");
+  if (cuda_dict == nullptr) {
+    throw TypeError("attribute `__cuda_array_interface__` must exist");
+  }
+
+  // In the following code block, we extract the values of `__cuda_array_interface__`
+  std::vector<int64_t> sizes;
+  ScalarType dtype;
+  npy_intp data_ptr;
+  std::vector<int64_t> strides;
+  try {
+    if (!PyDict_Check(cuda_dict)) {
+      throw TypeError("`__cuda_array_interface__` must be a dict");
+    }
+    PyObject *py_shape = PyDict_GetItemString(cuda_dict, "shape");
+    PyObject *py_typestr = PyDict_GetItemString(cuda_dict, "typestr");
+    PyObject *py_data = PyDict_GetItemString(cuda_dict, "data");
+    if (py_shape == nullptr || py_typestr == nullptr || py_data == nullptr) {
+      throw TypeError("attributes `shape`, `typestr`, and `data` must exist");
+    }
+
+    // Extract the `shape` attribute
+    int ndim = PySequence_Length(py_shape);
+    if (ndim == -1) {
+      throw TypeError("shape must be a sequence");
+    }
+    std::vector<npy_intp> npy_shape(ndim);
+    PyArray_IntpFromSequence(py_shape, npy_shape.data(), ndim);
+    sizes = to_aten_shape(ndim, npy_shape.data());
+
+    // Extract the `typestr` attribute
+    int dtype_size_in_bytes;
+    {
+      PyArray_Descr *descr;
+      if(!PyArray_DescrConverter(py_typestr, &descr)) {
+        throw ValueError("cannot parse `typestr`");
+      }
+      dtype = numpy_dtype_to_aten(descr->type_num);
+      dtype_size_in_bytes = descr->elsize;
+      assert(dtype_size_in_bytes > 0);
```

Comment:
Because `assert` cause a terminate() on failure, we don't use them in PyTorch codebase.  Use `TORCH_INTERNAL_ASSERT` to specify that something should be true (and a user should report a bug if it's not true.)

Reply:
Fixed in 5b0c83c